main.py

```diff
@@ -45,109 +68,71 @@ async def get_metadata():
-
+def validate_quota(quota: models.Quota):
     return {
-        "limit": quota.get("limit"),
-        "used": quota.get("used", None),
-        "type": quota.get("type", "token"),  # Default type to "token" if missing
-        "scope": scope,
-        "feature": quota.get("feature", None),
-        "user_id": quota.get("user_id", None)
+        "limit": quota.limit,
+        "used": quota.used,
+        "type": quota.type,
+        "scope": quota.scope,
+        "feature": quota.feature,
+        "user_id": quota.user_id
     }
 
 
 # Endpoint to get quotas (API Key protected)
 @app.get("/quota", response_model=List[QuotaGet], dependencies=[Depends(verify_api_key)])
-async def get_quotas():
-    try:
-        quotas = get_all_quotas()
-        validated_quotas = [validate_quota(q) for q in quotas]
-        return validated_quotas
-    except ValueError as e:
-        logging.error(f"Validation error: {e}")
-        raise HTTPException(status_code=422, detail=str(e))
+async def get_quotas(db: Session = Depends(get_db)):
+    quotas = crud.get_global_quotas(db)
+    validated_quotas = [validate_quota(q) for q in quotas]
+    return validated_quotas
 
 
 # Endpoint to update quotas (API Key protected)
 @app.put("/quota", dependencies=[Depends(verify_api_key)])
-async def put_quotas(quotas: List[QuotaUpdate]):
-    all_quotas = get_all_quotas()
+async def put_quotas(quotas: List[QuotaUpdate], db: Session = Depends(get_db)):
+    # Update passed quotas
     for quota in quotas:
-        quota_dict = quota.dict()
-        quota_dict["used"] = 0  # Initialize used to 0 for new quotas
-        quota_dict["type"] = quota_dict.get("type", "token")  # Default type
-        all_quotas.append(quota_dict)
-    update_quotas(all_quotas)
-    return {"message": "Quotas updated successfully"}
+        crud.update_or_create_global_quota(db, quota)
+
+    global_quotas = crud.get_global_quotas(db)
+    validated_quotas = [validate_quota(q) for q in global_quotas]
+    return validated_quotas
 
 
 # Endpoint to get quota for a specific course with user quotas option (API Key protected)
 @app.get("/quota/course/{course_id}", response_model=List[QuotaGet], dependencies=[Depends(verify_api_key)])
-async def get_course_quota(course_id: str, with_user_quotas: bool = Query(default=False)):
-    quotas = get_all_quotas()
-    course_quotas = [validate_quota(q) for q in quotas if
-                     q['scope'] in ['course', 'course-user'] and q.get('course_id') == course_id]
-    if not course_quotas:
-        raise HTTPException(status_code=404, detail="Course quota not found")
-    if with_user_quotas:
-        user_quotas = [validate_quota(q) for q in quotas if
-                       q['scope'] == 'course-user' and q.get('course_id') == course_id]
-        course_quotas.extend(user_quotas)
-    return course_quotas
+async def get_course_quota(course_id: str, db: Session = Depends(get_db)):
```

Comment:
did we get rid of the with_user_quotas parameter? Do we still need it?

Reply:
The user quota parameter was replaced by the route `/quota/course/{course_id}/user` in the specification which provides the same function.

Reply:
Ah, understood. Then I think everything is fine. I'll approve the merge.